x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,436 @@
+			if err != nil {
+				p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+			}
+			p.logger.Infof("Message visibility timeout updated to %v", visibilityTimeout)
+		}
+	}
+}
+
+func changeVisibilityTimeout(queueURL string, visibilityTimeout int64, svc *sqs.Client, receiptHandle *string) error {
+	req := svc.ChangeMessageVisibilityRequest(&sqs.ChangeMessageVisibilityInput{
+		QueueUrl:          &queueURL,
+		VisibilityTimeout: &visibilityTimeout,
+		ReceiptHandle:     receiptHandle,
+	})
+	_, err := req.Send(context.Background())
+	return err
+}
+
+func getRegionFromQueueURL(queueURL string) (string, error) {
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+}
+
+// handle message
+func handleSQSMessage(m sqs.Message) ([]s3Info, error) {
+	msg := sqsMessage{}
+	err := json.Unmarshal([]byte(*m.Body), &msg)
+	if err != nil {
+		return nil, errors.Wrap(err, "json unmarshal sqs message body failed")
+	}
+
+	var s3Infos []s3Info
+	for _, record := range msg.Records {
+		if record.EventSource == "aws:s3" && record.EventName == "ObjectCreated:Put" {
+			s3Infos = append(s3Infos, s3Info{
+				region: record.AwsRegion,
+				name:   record.S3.Bucket.Name,
+				key:    record.S3.Object.Key,
+				arn:    record.S3.Bucket.Arn,
+			})
+		}
+	}
+	return s3Infos, nil
+}
+
+func (p *Input) handleS3Objects(svc s3iface.ClientAPI, s3Infos []s3Info, errC chan error) {
+	if len(s3Infos) > 0 {
+		var wg sync.WaitGroup
+		wg.Add(len(s3Infos))
+
+		for i := range s3Infos {
+			go func(s3Info s3Info) {
+				defer wg.Done()
+				objectHash := s3ObjectHash(s3Info)
+
+				// read from s3 object
+				reader, err := bufferedIORead(svc, s3Info)
+				if err != nil {
+					errC <- errors.Wrap(err, "bufferedIORead failed")
+					return
+				}
+
+				offset := 0
+				for {
+					log, err := reader.ReadString('\n')
+					if log == "" {
+						break
+					}
+
+					if err != nil {
+						if err == io.EOF {
+							// create event for last line
+							offset += len([]byte(log))
+							err = p.forwardEvent(createEvent(log, offset, s3Info, objectHash))
+							if err != nil {
+								errC <- errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+							}
+							return
+						}
+
+						errC <- errors.Wrapf(err, "ReadString failed for %v", s3Info.key)
+						return
+					}
+
+					// create event per log line
+					offset += len([]byte(log))
+					err = p.forwardEvent(createEvent(log, offset, s3Info, objectHash))
+					if err != nil {
+						errC <- errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+						return
+					}
+				}
+			}(s3Infos[i])
+		}
+		wg.Wait()
+	}
```

Comment:
We span a number of go-routines here, all signaling errors to processKeepAlive. This looks like we end up with a number of zombie go-routines here. The problem is that processKeepAlive returns on the first event on errC, yet we have a many go-routines trying to send on a channel without a reader.

btw. assuming forwardEvent is asynchronous one can use a ref counting pattern like this (together with ACKEvents, which would call `c.Done()`):

```

type s3Context struct {
  mux sync.Mutex
  refs uint
  err error // first error witnessed or multi error
  errC chan error
}

func (c *s3Context) Inc() {
  c.mux.Lock()
  defer c.mux.Unlock()
  c.refs++
}

func (c *s3Context) Done() {
  c.mux.Lock()
  defer c.mux.Unlock()
  c.done()
}

func (c *s3Context) Fail(err error) {
  c.mux.Lock()
  defer c.mux.Unlock()

  // add "Typed" error to error

  c.done()
}

func (c *s3Context) done() {
  c.refs--
  if c.refs == 0 {
    c.errC <- c.err
    close(c.errC)
  }
}

    ...
    s3Context := &s3Context{
        refs: len(s3Infos),
        errC: errC
    }
    ...

    for _, info := range s3Info {
        ...
        go func() {
            defer s3Context.Done()

            ...
            s3Context.Inc() // we increase the ref count for each message. ACKEvent calls Done() to decrease it. -> The total ref count is number of messages + number of go-routines. errC will receive it's signal only after all go-routines have been stopped + all messages have been ACKed
            event := beat.Event{ // for each 
                Private: s3Context
           }
        }()
    }
```

We have a similar pattern in other places. Either with a function call, or with channel. If I find time, I will add this pattern (plus others?) to libbeat for reuse.

About typed errors: I wonder if it makes sense to introduce a typed error, that can update the message in SQS, such that already published events will not be send again later.

Reply:
Thank you @urso for the suggestion! Quick question, 
```
s3Context := &s3Context{
        refs: len(s3Infos),
        errC: errC
    }
``` 
For s3Context, we already set the refs to be `len(s3Infos)` to start, do we still need `s3Context.Inc()`? 

I added `s3Context` in commit https://github.com/elastic/beats/pull/12640/commits/a48b96f2a5c8ae78634504c011b8e503c109eb93 but removed it later. Because the length of s3Info is always 1 but aws documentation did say they might change that in the future. That's why I'm keeping `s3Infos` as an array but removed all goroutines here in `handleS3Objects` function. If things changed in the future, we decide goroutines are needed here again, then I will put `s3Context` back. Is that OK for you?

Reply:
Yes, you still need `s3Context.Inc`. The s3Context gets attached to the events private fields, meaning that the event itself becomes an owner of the context as well. The counter must only become 0 once there is no more owner, meaning no in-flight event or active worker producing new events and adding ownership to the event.

> I added s3Context in commit a48b96f but removed it later ... Is that OK for you?

Sure. If you don't need it, dont add it :)

Reply:
Turned out we still need the `s3Context` for implementing ACKEvents so I added back in 😬